cluster-autoscaler/core/scale_down.go

```diff
@@ -103,6 +103,93 @@ func (n *NodeDeleteStatus) SetDeleteInProgress(status bool) {
 	n.deleteInProgress = status
 }
 
+type resourceName string
+type scaleDownResourcesLimits map[resourceName]int64
+type scaleDownResourcesDelta map[resourceName]int64
+
+func computeScaleDownResourcesLeftLimits(nodes []*apiv1.Node, cp cloudprovider.CloudProvider, timestamp time.Time) (scaleDownResourcesLimits, errors.AutoscalerError) {
+	resourceLimiter, err := cp.GetResourceLimiter()
+	if err != nil {
+		return scaleDownResourcesLimits{}, errors.ToAutoscalerError(
+			errors.CloudProviderError,
+			err)
+	}
+
+	totals := make(map[resourceName]int64)
+	totals[cloudprovider.ResourceNameCores], totals[cloudprovider.ResourceNameMemory] = calculateCoresAndMemoryTotal(nodes, timestamp)
+
+	for gpuName, gpuCount := range calculateGpusTotal(nodes, cp, timestamp) {
+		totals[resourceName(gpuName)] = gpuCount
+	}
+
+	resultScaleDownLimits := make(scaleDownResourcesLimits)
+	for _, resource := range resourceLimiter.GetResources() {
+		total := totals[resourceName(resource)]
+		min := resourceLimiter.GetMin(resource)
+
+		// we put only actual limits into final map. No entry means no limit.
+		if min > 0 {
+			resultScaleDownLimits[resourceName(resource)] = total - min
+		}
+	}
+	return resultScaleDownLimits, nil
+}
+
+func noLimitsOnResources() scaleDownResourcesLimits {
+	return nil
+}
+
+func copyScaleDownResourcesLimits(source scaleDownResourcesLimits) scaleDownResourcesLimits {
+	copy := scaleDownResourcesLimits{}
+	for k, v := range source {
+		copy[k] = v
+	}
+	return copy
+}
+func computeScaleDownResourcesDelta(node *apiv1.Node, nodeGroup cloudprovider.NodeGroup) (scaleDownResourcesDelta, errors.AutoscalerError) {
+	nodeCPU, nodeMemory, err := getNodeCoresAndMemory(node)
+	if err != nil {
+		return scaleDownResourcesDelta{}, errors.ToAutoscalerError(errors.ApiCallError, err)
+	}
+
+	gpuType, gpuCount, err := gpu.GetNodeTargetGpus(node, nodeGroup)
+	if err != nil {
+		return scaleDownResourcesDelta{}, errors.ToAutoscalerError(errors.ApiCallError, err).AddPrefix("Failed to get node %v gpu: %v", node.Name)
+	}
+
+	return scaleDownResourcesDelta{
+		cloudprovider.ResourceNameCores:  nodeCPU,
+		cloudprovider.ResourceNameMemory: nodeMemory,
+		resourceName(gpuType):            gpuCount,
+	}, nil
+}
+
+func (limits *scaleDownResourcesLimits) checkDeltaWithinLimits(delta scaleDownResourcesDelta) *resourceName {
```

Comment:
Why return a pointer? I thought in go strings are already immutable and passed without copy?

Reply:
Also based on name I'd expect this function to return a bool...

Reply:
The semantics is to return name of resource which fails the limit, or nil if everything went fine. 
So pointer is only to have special "nil" value.
I considered returning two values (bool,  string). But found it to verbose.